### PR TITLE
VcsHost: Add special handling for URLs to the GitHub npm registry

### PR DESCRIPTION
**Pull request by oheger-bosch**
Such URLs are interpreted as VCS URLs of type GITHUB (based on their
host name), but they must not be processed like regular GitHub
repository URLs, as this would yield incorrect results.

This PR partly solves #5329. It makes sure that the URL is not changed in an incorrect way. There is, however, still a problem with the download in the scanner, because the deflation of the archive fails due to the missing file extension.

### CODE REVIEW

File: downloader/src/main/kotlin/VcsHost.kt [hunk 1]
```diff
@@ -158,29 +158,33 @@ enum class VcsHost(
             projectUrlToUserOrOrgAndProject(projectUrl)?.second?.removeSuffix(".git")
 
         override fun toVcsInfoInternal(projectUrl: URI) =
-            gitProjectUrlToVcsInfo(projectUrl) { baseUrl, pathIterator ->
-                var revision = ""
-                var path = ""
+            if (isGitHubNpmRegistryUrl(projectUrl.toString()))
```
[the remark this answers is not in the document]
Comment (oheger-bosch) [reply downloader/src/main/kotlin/VcsHost.kt:161]: Yes, I was also thinking about this; this is probably a cleaner solution. I will give it a try.

File: utils/common/src/main/kotlin/ArchiveUtils.kt [hunk 2]
```diff
@@ -87,8 +87,12 @@ fun archive(inputDir: File, zipFile: File, prefix: String = ""): File =
 /**
  * Unpack the [File] to [targetDirectory] using [filter] to select only the entries of interest.
  */
-fun File.unpack(targetDirectory: File, filter: (ArchiveEntry) -> Boolean = { true }) =
-    when (ArchiveType.getType(name)) {
+fun File.unpack(
```
Comment (sschuberth) [inline utils/common/src/main/kotlin/ArchiveUtils.kt:90]: Commit message nits:
- No comma in "To support cases, in which"
- "there is valid file name" -> "there is *a* valid file name"
- "to derive the archive type" -> "to derive the archive type *from*"

Maybe better after all: "To support cases where no archive type can be derived from the file name, allow to pass an explicit archive type fallback."

File: utils/common/src/main/kotlin/ArchiveUtils.kt [hunk 3]
```diff
@@ -87,8 +87,12 @@ fun archive(inputDir: File, zipFile: File, prefix: String = ""): File =
...
+    targetDirectory: File,
+    fallbackArchiveType: ArchiveType = ArchiveType.NONE,
+    filter: (ArchiveEntry) -> Boolean = { true }
+) =
+    when (ArchiveType.getType(name).takeUnless { it == ArchiveType.NONE } ?: fallbackArchiveType) {
```
Comment (sschuberth) [inline utils/common/src/main/kotlin/ArchiveUtils.kt:95]: I wonder whether we should really implement this as a fallback, or rather an override. An override would have the advantage that we could unpack wrong named archives. While that rarely happens, it seems like the more generic approach to always use the `ArchiveType` unless it is `NONE`. What do you think?

File: utils/common/src/main/kotlin/ArchiveUtils.kt [hunk 4]
```diff
@@ -163,7 +167,7 @@ fun File.unpackDeb(targetDirectory: File, filter: (ArchiveEntry) -> Boolean = {
             subDirectory.safeMkdirs()
 
             val file = tempDir.resolve(name)
-            file.unpack(subDirectory, filter)
+            file.unpack(subDirectory, filter = filter)
```
Comment (sschuberth) [inline utils/common/src/main/kotlin/ArchiveUtils.kt:170]: Unnecessary change.

File: downloader/src/main/kotlin/Downloader.kt [hunk 5]
```diff
@@ -328,7 +329,7 @@ class Downloader(private val config: DownloaderConfiguration) {
                     gemDirectory.safeDeleteRecursively(force = true)
                 }
             } else {
-                sourceArchive.unpack(outputDirectory)
+                sourceArchive.unpack(outputDirectory, fallbackArchiveType = ArchiveType.TAR_GZIP)
```
Comment (sschuberth) [inline downloader/src/main/kotlin/Downloader.kt:332]: I don't feel comfortable with hard-coding this at this central place. I haven't looked at the details of the issue you're trying to address, but maybe we could instead set a correct file extension based on a mime-type in a HTTP response?

Comment (sschuberth) [reply downloader/src/main/kotlin/Downloader.kt:332]: I have some more ideas. Can we maybe have a look at this together?

File: utils/common/src/main/kotlin/ArchiveUtils.kt (hunk 2 above)
Comment (oheger-bosch) [reply utils/common/src/main/kotlin/ArchiveUtils.kt:92]: Reworded and adapted to the override logic.

File: utils/common/src/main/kotlin/ArchiveUtils.kt (hunk 3 above)
Comment (oheger-bosch) [reply utils/common/src/main/kotlin/ArchiveUtils.kt]: Yes, this is better. Switched to override logic.

File: utils/common/src/main/kotlin/ArchiveUtils.kt (hunk 4 above)
Comment (oheger-bosch) [reply utils/common/src/main/kotlin/ArchiveUtils.kt:172]: This is actually needed because of the new optional parameter in `unpack()`.

File: downloader/src/main/kotlin/Downloader.kt (hunk 5 above)
Comment (oheger-bosch) [reply downloader/src/main/kotlin/Downloader.kt]: Sure. I have pushed my current state.

File: utils/common/src/main/kotlin/ArchiveUtils.kt [hunk 6]
```diff
@@ -85,10 +85,16 @@ fun archive(inputDir: File, zipFile: File, prefix: String = ""): File =
     }
 
 /**
- * Unpack the [File] to [targetDirectory] using [filter] to select only the entries of interest.
+ * Unpack the [File] to [targetDirectory] using [filter] to select only the entries of interest. If aan archive type
```
Comment (sschuberth) [inline utils/common/src/main/kotlin/ArchiveUtils.kt:88]: Typo: "aan" -> "an"

File: utils/common/src/main/kotlin/ArchiveUtils.kt [hunk 7]
```diff
@@ -85,10 +85,16 @@ fun archive(inputDir: File, zipFile: File, prefix: String = ""): File =
...
+ * other than [ArchiveType.NONE] is specified in [useArchiveType], use this one; otherwise, try to detect the type
+ * based on the file extension.
  */
-fun File.unpack(targetDirectory: File, filter: (ArchiveEntry) -> Boolean = { true }) =
-    when (ArchiveType.getType(name)) {
+fun File.unpack(
+    targetDirectory: File,
+    useArchiveType: ArchiveType = ArchiveType.NONE,
```
Comment (sschuberth) [inline utils/common/src/main/kotlin/ArchiveUtils.kt:94]: Nit: Maybe better `forceArchiveType`?

**Review by sschuberth — CHANGES_REQUESTED** [review]
